Python/chapter01/1.8 - Zero Matrix/miguel_1.8_sol.py

```diff
@@ -0,0 +1,179 @@
+                ],
+                [
+                    [1, 2, 0, 4],
+                    [0, 0, 0, 0],
+                    [9, 10, 0, 12],
+                    [13, 14, 0, 16]
+                ]
+            ),
+            (
+                [
+                    [1, 2],
+                    [3, 4],
+                    [5, 6],
+                    [7, 8],
+                    [9, 0]
+                ],
+                [
+                    [1, 0],
+                    [3, 0],
+                    [5, 0],
+                    [7, 0],
+                    [0, 0]
+                ],
+            ),
+            (
+                [
+                    [1, 2],
+                    [0, 4],
+                    [5, 6],
+                    [7, 8],
+                    [9, 0]
+                ],
+                [
+                    [0, 0],
+                    [0, 0],
+                    [0, 0],
+                    [0, 0],
+                    [0, 0]
+                ]
+            ),
+            (
+                [
+                    [1, 2, 3, 4, 5, 6, 0, 8, 9],
+                    [9, 8, 7, 6, 5, 4, 3, 2, 1],
+                ],
+                [
+                    [0, 0, 0, 0, 0, 0, 0, 0, 0],
+                    [9, 8, 7, 6, 5, 4, 0, 2, 1],
+                ]
+            ),
+            (
+                [
+                    [0]
+                ],
+                [
+                    [0]
+                ]
+            ),
+            (
+                [
+                    [1]
+                ],
+                [
+                    [1]
+                ]
+            ),
+            (
+                [
+                    [1, 2],
+                    [3, 4]
+                ],
+                [
+                    [1, 2],
+                    [3, 4]
+                ]
+            ),
+            (
+                [
+                    [0, 1, 2],
+                    [0, 3, 4],
+                    [0, 5, 6],
+                    [0, 7, 8]
+                ],
+                [
+                    [0, 0, 0],
+                    [0, 0, 0],
+                    [0, 0, 0],
+                    [0, 0, 0]
+                ]
+            ),
+            (
+                [
+                    [1, 2, 3, 4],
+                    [5, 6, 0, 8],
+                    [9, 10, 0, 12],
+                    [13, 14, 15, 16]
+                ],
+                [
+                    [1, 2, 0, 4],
+                    [0, 0, 0, 0],
+                    [0, 0, 0, 0],
+                    [13, 14, 0, 16]
+                ]
+            ),
+            (
+                [
+                    [0, 0, 0, 0],
+                    [1, 2, 3, 4],
+                    [5, 6, 7, 8],
+                    [9, 10, 11, 12]
+                ],
+                [
+                    [0, 0, 0, 0],
+                    [0, 0, 0, 0],
+                    [0, 0, 0, 0],
+                    [0, 0, 0, 0]
+                ]
```

Comment:
can you add

```python
[
    [[0] * 1000 for _ in range(1000)],
    [[0] * 1000 for _ in range(1000)],
],
```

? a O(MN) algorithm should have no trouble with that.

Reply:
After adding that test case, the code ran and passed the test cases in about 112 seconds.  Very slow.  In the worst case (all 0s in input matrix), would the runtime be O(M*N * (M + N)) = O(M^2*N + N^2*M)?  Because in that case of all 0s input, each element in the matrix will do M + N work.

My solution is very slow. I'm going to try to optimize it.

Reply:
Updated